### PR TITLE
Add DNSRecord controller

### PR DESCRIPTION
**Pull request by stoyanr**
**How to categorize this PR?**

/area control-plane
/kind api-change
/kind enhancement
/platform aws

**What this PR does / why we need it**:
* Adds a AWS-specific implementation of a DNSRecord controller, see https://github.com/gardener/gardener/pull/4231.
* Vendors g/g v1.26.0-dev and k8s v0.21.1 (in a separate commit to make reviews easier)

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
Note the following important details:
* The provider type is `aws-route53` (like in `DNSProvider`), not `aws` (like in other controllers) for reasons of compatibility and ease of migration.
* The credentials secret keys are again the ones compatible with existing domain secrets, e.g. `AWS_ACCESS_KEY_ID` instead of `accessKeyID`.
* The functions `CreateOrUpdateDNSRecord` and `DeleteDNSRecord` return after the DNS record is created / updated / deleted and don't wait until the change has been synced to all nameservers. (/cc @MartinWeindel)

I tested creating, updating, and deleting DNS records with the controller locally and it seems to work as expected.

**Release note**:

```other operator
NONE
```

### CODE REVIEW

**Comment by MartinWeindel** [comment]
I don't understand the rationale of this PR at all. I'm surprised to see the DNS functionality apparently distributed to multiple extensions. 
1. DNS server is now linked with the infrastructure, although there is no direct dependency between specification of DNS provider and infrastructure. Should the aws-provider now be deployed on all seeds, if the DNS is hosted on AWS? How do you deal if a Gardener installation uses netlify, cloudflare or infoblox as DNS provider? Do you want to create new infrastructure providers for this purpose? What happens if internal and external DNS domains are on different backends?
Additionally the Gardenlet now needs to watch what DNS providers it needs on a seed and needs to make sure, that the corresponding infrastructure controller is deployed as soon as the a hosted zone is used from DNS provider type which is not deployed currently.  
2. I don't understand how the `DNSProvider` can be dropped without creating a similar new specification. Even if by chance DNS domain and infrastructure are using the same backend/SDK, they are using **different accounts** practically all the time. This means you need different secrets for infrastructure and DNS hosted zone management. As the same DNSProvider is often used for the `external` DNS entry and services/ingresses via shoot-dns-service, you have a duplication of the secrets.
3. "DNS hosted zones are cached": This is ok for a draft, but this is not sufficient for production purposes. You have to use a cache with TTL and renew the cache periodically. Especially if end users use their own zones, you need to be aware of such new zones in a short time. 
4. "`spec.dns.domain` can't be managed by a custom provider anymore" Is this really your target? If yes, how do you deal with all the clusters using their own hosted zones for the external and ingress DNS currently?

**Comment by stoyanr** [comment]
@MartinWeindel Thanks for your feedback! The design is a direct consequence of the proposal we discussed. To your questions.

> Should the aws-provider now be deployed on all seeds, if the DNS is hosted on AWS? 

Yes. We might optimize it so that only the DNS controller is started in this case. This can be done later as it's pure configuration.

> How do you deal if a Gardener installation uses netlify, cloudflare or infoblox as DNS provider? Do you want to create new infrastructure providers for this purpose? 

Yes, they will only contain a DNS controller.

> What happens if internal and external DNS domains are on different backends?

We will have both extensions installed on the corresponding seed.

> Additionally the Gardenlet now needs to watch what DNS providers it needs on a seed and needs to make sure, that the corresponding infrastructure controller is deployed as soon as the a hosted zone is used from DNS provider type which is not deployed currently.

Yes, and this is exactly the mechanism used for all other extensions (provider, runtime, network, etc.), and so we'll actually be happy to use it for DNS as well.

> I don't understand how the DNSProvider can be dropped without creating a similar new specification. Even if by chance DNS domain and infrastructure are using the same backend/SDK, they are using different accounts practically all the time. This means you need different secrets for infrastructure and DNS hosted zone management. As the same DNSProvider is often used for the external DNS entry and services/ingresses via shoot-dns-service, you have a duplication of the secrets.

We don't want the complication of DNS provider, in fact getting rid of it is one of our main motivations for this change. If different accounts are used, we will have different secrets for them, exactly as today. If the same DNS provider is used as you mention above (and only in this case), then we would have the same credentials in 2 different secrets - one created for the DNSProvider, and another one for the DNSRecord (so 1 secret more than today). We don't see this as much of an issue.

> "DNS hosted zones are cached": This is ok for a draft, but this is not sufficient for production purposes. 

In fact, we think it's better to get rid of caching entirely and minimize the need for getting the DNS hosted zones from route53 via other means. DNSRecords will only be reconciled when the shoot is created and when it's migrated to a different seed (similarly to other extension controllers, it uses the `gardener.cloud/operation` annotation to implement this restriction). Since the hosted zone never changes, once it's written to the status during the first reconciliation it can be used directly, which leaves us with one single getting of hosted zones per shoot for the shoot's lifetime. To minimize this even further, we will add the zone as an optional field to the DNSRecord spec and actually pass it for our own hosted zones. With all that, we believe the need for caching goes away entirely.

> "spec.dns.domain can't be managed by a custom provider anymore" Is this really your target? If yes, how do you deal with all the clusters using their own hosted zones for the external and ingress DNS currently?

I was thinking about deprecating `spec.dns.providers.primary: true` in favor of fields like `spec.dns.type`, `spec.dns.region`, `spec.dns.zone`, etc., similarly to e.g. `network` or `backup` - so it's still possible to use a user-specified type, zone, etc. for Gardener DNS records, but not via `spec.dns.providers`. We still see this as an option for a future version of the API, however for now we would like to avoid incompatible changes, so we will allow `spec.dns.providers.primary: true` and simply ignore the `domains` and `zones` in this case for Gardener in this case.

/cc @rfranzke @timebertt @plkokanov

**Comment by stoyanr** [comment]
Corrected the PR (and also https://github.com/gardener/gardener/pull/4231) to reflect the discussion above, /cc @timebertt @MartinWeindel if you would like to take another look.

**Review by MartinWeindel — CHANGES_REQUESTED** [review]
Found one problem in the findZoneForName function.
Additiionally, alias targets should also be supported for AWS to reduce costs.
See https://github.com/gardener/external-dns-management/issues/27
and [code for creating alias target records instead of A records](https://github.com/gardener/external-dns-management/blob/master/pkg/controller/provider/aws/aliastarget.go#L98-L115)

**Comment by MartinWeindel** [comment]
> In fact, we think it's better to get rid of caching entirely and minimize the need for getting the DNS hosted zones from route53 via other means. DNSRecords will only be reconciled when the shoot is created and when it's migrated to a different seed (similarly to other extension controllers, it uses the gardener.cloud/operation annotation to implement this restriction). Since the hosted zone never changes, once it's written to the status during the first reconciliation it can be used directly, which leaves us with one single getting of hosted zones per shoot for the shoot's lifetime. To minimize this even further, we will add the zone as an optional field to the DNSRecord spec and actually pass it for our own hosted zones. With all that, we believe the need for caching goes away entirely.

Good idea

File: pkg/aws/client/client.go
Comment (MartinWeindel) [inline pkg/aws/client/client.go:483]: Load balancers in AWS are using domain names instead of IP addresses. Therefore `CNAME` records need to be supported (and converted into `ALIAS` target records typically). But as far as I can see, the implementation already supports it, as the record type is not validated. `CNAME` records can only have a single value, but an explicit validation is probably not needed as the AWS route53 service will report an error.

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go:142]: It will not (`strings.HasSuffix("d.c.b.a", "ccc.b.a") == false`). You probably mean "d.cccc.b.a" must not match "ccc.b.a", in which case I agree. Will use the propose matching function, thanks!

File: pkg/aws/client/client.go
Comment (stoyanr) [reply pkg/aws/client/client.go:483]: Thanks a lot! So if I understand correctly, in this case:
* The `RecordType` of the DNSRecord in this case should be `CNAME`. We'll add explicit validation for only a single value in this case (as well as for other things, e.g. the supported values for `recordType`) to https://github.com/gardener/gardener/pull/4231 before merging it.
* Only for the `aws-route53` implementation, this can be converted to an "alias target" using the code you referred to above as an example. 

Question: What is the advantage of using "alias target" instead of a regular `CNAME`? Could we simply live with `CNAME` everywhere?

Comment (stoyanr) [reply pkg/aws/client/client.go:483]: OK, I found https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resource-record-sets-choosing-alias-non-alias.html and this somewhat answers my last question. Perhaps the important one is "Pricing for DNS queries" (e.g. alias records are "free" and CNAME are not). Are there any other considerations?

Can the hardcoded list https://github.com/gardener/external-dns-management/blob/2b72ef8cb031137d2945b8253a27dcb1042e733d/pkg/controller/provider/aws/aliastarget.go#L29 be avoided? How likely is it change and to have to be actively maintained?

Comment (MartinWeindel) [reply pkg/aws/client/client.go:483]: You guessed right. The usage of alias was requested from the SAP Hana team to reduce costs for DNS queries.

We have found no other solution than to hardcode the list. There seems to be no API to fetch this list. We even considered scraping HTML from the page [Elastic Load Balancing endpoints and quotas](https://docs.aws.amazon.com/general/latest/gr/elb.html), but this is of course not robust. The assumption is that the zone IDs never change. If there are new AWS regions we may not be able to generate `ALIAS` as long as they are not appended to the list.

Comment (stoyanr) [reply pkg/aws/client/client.go:483]: OK, thanks, will do as suggested then.

Comment (stoyanr) [reply pkg/aws/client/client.go]: Done.

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go]: Done.

File: pkg/controller/dnsrecord/actuator.go
Comment (timebertt) [inline pkg/controller/dnsrecord/actuator.go:125]: We probably want to have some ref to the dnsrecord object in all of those logs.

**Review by timebertt — COMMENTED** [review]
Well, there is not much to comment on. Looks very good already :)
I didn't pay too much attention to the DNS details, as I'm not so experienced in this area as @MartinWeindel. The controller structure and so on looks very good.

Will play around with it and test it a bit next.

File: pkg/aws/secret.go
Comment (stoyanr) [reply pkg/aws/secret.go:47]: I could do it, but I would like to be sure that there is a use case for this. If gardener would never pass DNS secrets in the second vairant, then the added complexity is not worth it.

> I remember, there is a use case in gardener, where you specify a non-default domain in Shoot.spec.dns.domain but no custom dns provider (secret). In that case the shoots credentials are used. 

If I specify a non-default domain while having a `default-domain` secret I get an error during validation:
```
Error from server (BadRequest): error when creating "dev/shoot-i500152-gcp.yaml": shoot uses a default domain but does not match expected scheme: <shoot-name>.<project-name>.<default-domain>
```

And if I delete the `default-domain` secret, the shoot is accepted, but the reconciliation fails immediately with:
```
Could not initialize a new operation for Shoot reconciliation: unable to figure out which secret should be used for dns.
```

Without a `default-domain` secret, the only way I could reconcile such a shoot I found so far is to add a provider (with a secret) and set `primary` to `true`. So I don't think what you describe is actually possible. Could I be missing something?

Having in mind that:
* `external-dns-management` would so far only accept the "DNS format" of secrets. 
* gardener doesn't perform any mapping of secret keys (and I think it shouldn't, since it's provider-specific). 

The only reason to accept the other secret format would be to allow users who never used `external-dns-management` before to specify a secret in this format for a custom provider. I am sceptical that this justifies the increased complexity of supporting both secret variants in all providers (this is just one of them).

Comment (stoyanr) [reply pkg/aws/secret.go:47]: Note: I double checked, `external-dns-management` indeed supports the "other format" via "alternative keys", e.g.:
```
accessKeyID, err := c.GetRequiredProperty("AWS_ACCESS_KEY_ID", "accessKeyID")
```
Still, the above consideration still applies. Is there a use case in which this format is actually used currently? If not, are we sure we want to introduce one?

File: pkg/aws/client/client.go
Comment (stoyanr) [reply pkg/aws/client/client.go:426]: Yes, you are right, will extract it.

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go:125]: Good point, will add it.

File: pkg/controller/dnsrecord/actuator.go
Comment (timebertt) [inline pkg/controller/dnsrecord/actuator.go:114]: This means, once the zone in the status is set, there is no way to influence it anymore via the spec field.

Just to be on the safe side:
Is there a use case where we might want to prefer the spec over the status field? I could imagine that if you deleted the record and recreated the hosted zone in between reconciliations, the controller would not be able to recover from this even if you specify the new hosted zone in the spec. 
Though, I'm not sure if that's entirely correct, so I'm asking here :)

File: pkg/controller/dnsrecord/actuator.go
Comment (timebertt) [inline pkg/controller/dnsrecord/actuator.go:50]: One thing I noticed during my tests:
The controller uses the default workqueue rate limiter, which means failures are retried very quickly.
E.g. when specifying the wrong zone ID, this will result in a bunch of requests in a few seconds.

We might want to
- a) use a rate limiter with different qps/baseDelay settings or
- b) set `RetryAfter` (or influence it somehow from the actuator) or
- c) try to detect such errors (e.g. reusing error code detection early on here in the actuator) and only retry a few times
- or a combination of those options

in order to not immediately exhaust the accounts rate limits on such configuration issues.

File: pkg/aws/secret.go
Comment (stoyanr) [reply pkg/aws/secret.go:47]: I double checked some more and found this code: https://github.com/gardener/gardener/blob/54930d430301a5f0aadfa5367e10012371c542d0/pkg/operation/shoot/shoot.go#L466-L475
Trying to test it now ...

File: pkg/controller/dnsrecord/actuator.go
Comment (timebertt) [inline pkg/controller/dnsrecord/actuator.go:78]: Ok, this might be a bit theoretical and I'm not sure, if we need to think about this case, but hear me out :)

If the record type changes for some reason (might be because the corresponding LB changes from IP to Hostname, or because the order in the values list is changed and now the first one isn't part of a canonical zone anymore, so we change from ALIAS to CNAME, or whatever else), we will orphan the old record in the infrastructure.

If this is indeed a problem, we might want to track the record type that was created in the infrastructure and accordingly a) deny switching the record type or b) delete the old entry before creating one of the new type, or something similar.

Actually, the same applies to a changing name field value (though this one should definitely not change, I suppose). So another option might be to make those fields immutable, though I'm not sure if that's easily achievable via CRDs or if we would need validating webhooks.

File: pkg/controller/dnsrecord/actuator.go
Comment (timebertt) [inline pkg/controller/dnsrecord/actuator.go:65]: I don't remember the details: is the region an important detail in Route53? If so, I might make sense to add a respective field to this log message.

I don't see anything about regions in the dashboard and its not "regionized", so I suppose it's not important. But I don't know the details about the gov cloud use case.

**Review by timebertt — COMMENTED** [review]
Ok, finished testing it. Worked really great!

A few more question arose while playing around with it, added them here to make sure we don't miss any details.

File: pkg/aws/secret.go
Comment (timebertt) [reply pkg/aws/secret.go:47]: > I double checked some more and found this code: https://github.com/gardener/gardener/blob/54930d430301a5f0aadfa5367e10012371c542d0/pkg/operation/shoot/shoot.go#L466-L475
> Trying to test it now ...

Yep, exactly. Also had to dig a bit, but this is the use case that I remembered.
So, a non-default domain is used in conjunction with a primary dns provider that doesn't reference a secret.

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go:114]: I think this is really minor, but it would be easy to just reverse the 2 cases and make sure the spec takes precedence if it's specified. I will do it.

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go:78]: I don't think this is what will happen if the record type changes. If you specify a record with the correct name and wrong record type, you will get an error from route53 different from "not found". So the record will not be orphaned, instead the deletion will fail. Perhaps this is good enough for a "theoretical" issue like this?

Regarding the name, this should definitely not change. It should be made immutable (and perhaps the record type as well). I will check what is the best way to do it and add it to https://github.com/gardener/gardener/pull/4231.

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go:65]: I also don't think the region is important. It's needed to create the client, but apart from that it doesn't seem to make a difference. I think the same records will be managed if the region changes. So I would avoid making it too prominent.

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go:50]: OK, good point. I think b) is easiest since we have `ReconcileErr` to override the `RequeueAfter` value returned by the reconciler. We should perhaps set this to e.g. 30 seconds for all errors that are coming from the route53 API, so no sophisticated error detection is needed? 

Regarding "only retry a few times" this would mean saving the number of retries in the status. I would avoid it unless you think it's really important.

Comment (timebertt) [reply pkg/controller/dnsrecord/actuator.go:50]: > OK, good point. I think b) is easiest since we have `ReconcileErr` to override the `RequeueAfter` value returned by the reconciler. We should perhaps set this to e.g. 30 seconds for all errors that are coming from the route53 API, so no sophisticated error detection is needed?

Hmmm, if we use a constant backoff of `30s` this might even make more calls in the long run that an exponential backoff, right?
But I'm not too sure about what's the best option here. We can take this to our in-person discussion tomorrow.

> Regarding "only retry a few times" this would mean saving the number of retries in the status. I would avoid it unless you think it's really important.

I can't really judge as of now, as I don't have enough hands-on experience with those things. Maybe @MartinWeindel can already give an assessment based on his experiences.
But in any case, we can still introduce such mechanism later on, if it becomes necessary.

File: pkg/controller/dnsrecord/actuator.go
Comment (timebertt) [reply pkg/controller/dnsrecord/actuator.go:65]: Ok, good. Then let's keep it like it is.

File: pkg/aws/secret.go
Comment (stoyanr) [reply pkg/aws/secret.go]: Done. Tested, works the use case mentioned above, works fine.

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go]: Done.

File: pkg/aws/client/client.go
Comment (stoyanr) [reply pkg/aws/client/client.go]: Done.

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go]: Done.

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go:60]: Implemented and tested it with a hardcoded requeueAfter of 30 seconds and a wrongly configured zone. The actual requests happen more often than 30 seconds in the beginning because gardenlet also retries. After gardenlet stops retrying, they continue happening every 30 seconds. Perhaps the value should be increased to 1 minute? I don't think it should be made configurable though, it's just a precaution against rate limits.

**Comment by stoyanr** [comment]
@timebertt I believe I addressed all your feedback (in a separate commit).
/squash

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go:60]: > Hmmm, if we use a constant backoff of 30s this might even make more calls in the long run that an exponential backoff, right?

Yes, but I think rate limits are per unit of time, so this doesn't really matter? The important thing is not to have too many requests quickly in succession, especially if many shoots are reconciled at the same time with a wrong configuration. I am not really sure either, but as I said I would not start with some elaborate mechanism at this point just to mitigate this issue.

BTW, I think gardenlet retries mentioned above could be prevented by correctly classifying such errors as user errors. Right now this is not the case, so I am seeing more requests than optimal. I guess 30 seconds is perhaps ok in this case.

**Review by timebertt — COMMENTED** [review]
Thanks for incorporating my suggestions!
Also no further feedback to this PR right now from my side :)

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go:60]: When thinking more about it, I think the one practical way to deal with the issue would be to combine b) and c) - detect configuration errors if at all possible (e.g. the error code `NoSuchHostedZone` can't be anything else but a configuration error), and for such errors return a really long `RequeueAfter`, e.g. 1 hour. For all other route53 errors, return e.g. 30 seconds or simply propagate the error as is.
We could also recognize rate limit errors (I believe this is the `ThrottlingException` code), and `RequeueAfter` 1 hour here as well (on errors), since if we are throttled it's obviously pointless to retry more often.

Comment (timebertt) [reply pkg/controller/dnsrecord/actuator.go:60]: True, all good what you say. 
However, I think it's enough to save these ideas for later, once we actually have the problems. The current approach is not the nicest, but is probably already enough.

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go:91]: > It seems the behavior depends on the record type ...

In fact I was wrong, the same happens when trying to switch from A to CNAME and vice versa - the update failed as I expected, but then the deletion of the DNSRecord succeeded (it failed on route53 but the error returned was "not found" and so ignored), and so the actual record in route53 was orphaned.

Unresolving this since it's not so trivial, we should at least give it some thought.

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go]: In fact, it seems the region is not used **at all**. I tried with regions `foo` and `bar` and everything still works ... Still, this is route53-specific behavior, so I believe we should keep the region in the API as it *might* be important on other infrastructures.

Comment (timebertt) [reply pkg/controller/dnsrecord/actuator.go]: Sounds good, thanks for double-checking.

Comment (MartinWeindel) [reply pkg/controller/dnsrecord/actuator.go]: The region **is used** in some special cases:
1. "Route 53 in the Beijing and Ningxia Regions: specify cn-northwest-1." see https://docs.aws.amazon.com/general/latest/gr/r53.html
2. we make use of regions  for NS2 (AWS regional API access). It also needs another endpoint: see https://github.com/gardener/external-dns-management/blob/master/pkg/controller/provider/aws/handler.go#L99-L101
/cc @danielfoehrKn

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go:60]: We discussed the issue and decided not to do any further changes. Instead, we should rely on gardener properly categorizing these errors as user errors via https://github.com/gardener/gardener/blob/677eaffc5acdf78d19f09566945f4e91d97da742/pkg/apis/core/v1beta1/helper/errors.go#L56-L64 (the regexes should be extended appropriately with the corresponding route53 errors, as well as those from other providers). In this case, the shoot will rather quickly become `Failed`, and our extension controller will stop reconciling its resources, which will prevent further requests to the API. In essence, we already have an appropriate mechanism to deal with this, and we should use it instead of introducing an a different one.

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go:91]: We discussed this issue and decided not to introduce more complexity at this stage, apart from making the `recordType` field also immutable and making sure that it's actually validated by Gardener.

**Comment by stoyanr** [comment]
Added unit tests in a new commit. I consider this PR ready now except for the fact that it's based on https://github.com/gardener/gardener/pull/4231, which is still not merged, so keeping the Draft state until this is the case.

File: pkg/controller/dnsrecord/actuator.go
Comment (timebertt) [inline pkg/controller/dnsrecord/actuator.go:177]: Nit
```suggestion
	case len(credentials.Region) > 0:
```

**Review by timebertt — COMMENTED** [review]
Codewise lgtm, will test once https://github.com/gardener/gardener/pull/4312 is merged and vendored.

File: pkg/controller/dnsrecord/actuator.go
Comment (stoyanr) [reply pkg/controller/dnsrecord/actuator.go]: Done

**Comment by stoyanr** [comment]
Now this is properly based on g/g and therefore can be merged if there is no further feedback. Removing `Draft`.

**Review by timebertt — APPROVED** [review]
Thank you very much!
Code-wise
/lgtm
now!

Won't been able to test it myself again, but my first tests were already very good and not much has changed since then.

File: pkg/aws/secret.go
Comment (rfranzke) [inline pkg/aws/secret.go:31]: Why "allow"DNSKeys?

**Review by rfranzke — APPROVED** [review]
/lgtm

File: pkg/aws/secret.go
Comment (stoyanr) [reply pkg/aws/secret.go:31]: Because they can be either allowed or not. If they are allowed, then a secret containing them will be parsed successfully. If not, parsing such a secret will fail. The idea is, that a secret that contains DNS keys but is not used for DNS will still fail to be parsed (we don't want to suddenly allow a new set of secret keys for everything).

Comment (rfranzke) [reply pkg/aws/secret.go:31]: Makes sense, thank you!

**Pull request merged**
